### PR TITLE
Convert all python code to python3. Fixes #12506.

### LINKED ISSUE #12506: Should switch to Python 3
Python 2.x will no longer be supported by their upstream developers in 2020.
Thus Debian developers are actively removing Python 2 support in Debian Testing with the goal of getting rid of Python 2 in Debian 11 (bullseye).

Kali is tracking Debian Testing and is thus affected by this.

You should consider to switch the Python 2 scripts to Python 3.

Thanks

### PR DESCRIPTION
**Pull request by xmunoz**
This is a breaking change that removes support for python2. Not every file in the change has been fully tested because of limited access to the exploit target environments, but running with `python3 $FILE` produced no syntax errors when the needed dependencies were available. These changes were made with the following command:

    2to3-2.7 -f all -f ws_comma -w -n $FILE

[2to3](https://docs.python.org/3/library/2to3.html) is the recommended utility for auto translation for python 2 -> 3. I also manually edited some files where 2to3 performed unexpectedly.

Related: https://gitlab.com/kalilinux/packages/metasploit-framework/issues/1

Fixes #12506.

Verification
========

- [x] When using Python 3 (may require temporarily modifying a symlink or the shebang in the module files to force)
  - [x] The windows shellcode builders produce the correct output
  - [x] The teradata modules work per https://github.com/rapid7/metasploit-framework/pull/10109
  - [x] The impacket modules work per #9816 #10106 #10107
  - [x] The haraka module works per #8178
  - [x] `exploit/windows/smb/ms17_010_eternalblue_win8` works per #10184
- [x] When using Python 2.6+ (may require temporarily modifying a symlink or the shebang in the module files to force)
  - [x] The teradata modules work per https://github.com/rapid7/metasploit-framework/pull/10109
  - [x] The impacket modules work per #9816 #10106 #10107
  - [x] The haraka module works per #8178
  - [x] `exploit/windows/smb/ms17_010_eternalblue_win8` works per #10184

### CODE REVIEW

**Comment by wvu-r7** [comment]
Wow, thanks!

File: data/exploits/cve-2015-5287/sosreport-rhel7.py [hunk 1]
```diff
@@ -48,40 +48,40 @@ def waitpath(p):
...
-print "tmpfiles: ", tmpfiles
+print("tmpfiles: ", tmpfiles)
 
-print "moving directory"
+print("moving directory")
 
 os.rename(sosreport, sosreport + ".old")
 os.mkdir(sosreport)
-os.chmod(sosreport,0777)
+os.chmod(sosreport,0o777)
```
Comment (acammack-r7) [inline data/exploits/cve-2015-5287/sosreport-rhel7.py]: This is an old exploit that runs only on targets that have Python 2 installed as the default and does not need to be updated.

**Review by acammack-r7 — CHANGES_REQUESTED** [review]
Thanks for your work on this! I think we will want a somewhat different approach from this, though. First, the output of 2to3 is a bit quirky and there are some things, like extra calls to `list()`, that we do not need or want. Next, a couple of these files are intended to run on older systems that definitely have Python 2 but may not have Python 3, or if the user has installed it, it may not be where we expect (also, the module that runs this privilege escalation scripts would need to be updated to look for Python 3). Finally, Python 2 is still around for a bit yet, and since dual-compatibility is not a huge lift for us I am reluctant to wholesale drop support for it in modules or external module libraries that do not explicitly need Python 3 features (like the async stuff). Tools and build scripts I think are fine to move to 3, since they are intended to be run by developers.

On the shebangs, files that can run under 2 and 3 should be allowed to do so. Also, as I understand it, the encoding can now taken from the operating system's `LOCALE` as well, which means we still have to specify the encoding to ensure the files are read properly.

Thanks again for submitting this, there are definitely a couple places you caught that I wouldn't have thought to check!

**Comment by xmunoz** [comment]
Thanks for the comprehensive review! I'll clean this up shortly.

File: data/exploits/cve-2015-5287/sosreport-rhel7.py (hunk 1 above)
Comment (xmunoz) [reply data/exploits/cve-2015-5287/sosreport-rhel7.py]: Ok, I'll revert all the changes to this file.

**Review by acammack-r7 — APPROVED** [review]

**Comment by acammack-r7** [comment]
Thanks for the cleanup, the changes look good! I haven't had a chance to test all the moving parts yet, but I have added some verification steps to the OP for anyone who wants to help out verifying these changes.

**Comment by xmunoz** [comment]
When I was testing the windows shellcode builders, I encountered 2 errors on master.

1) On `external/source/shellcode/windows/x64/build.py`:

    $ python build.py all
    ...
    ./src/block/block_reverse_https.asm:148: warning: `ptr' is not a NASM keyword [-w+ptr]
    ./src/block/block_reverse_https.asm:148: error: comma, colon, decorator or end of line expected after operand
    [-]  [Errno 2] No such file or directory: 'bin/stager_reverse_https.bin'

According to [this](https://stackoverflow.com/a/13791007/1299373), you can just remove the `ptr` keyword. When I did that, the script ran to completion. I ran it on master and this branch, and compared the output. The results were identical.
```
$ diff result-master.txt result-python3.txt
1c1
< # Built on Fri Nov  8 10:47:56 2019
---
> # Built on Fri Nov  8 10:51:53 2019
```
Should I add this change to this PR?

```
$ git diff external/source/shellcode/windows/x64/src/block/block_reverse_https.asm
diff --git a/external/source/shellcode/windows/x64/src/block/block_reverse_https.asm b/external/source/shellcode/windows/x64/src/block/block_reverse_https.asm
index 7578b590c1..604d47c8ad 100644
--- a/external/source/shellcode/windows/x64/src/block/block_reverse_https.asm
+++ b/external/source/shellcode/windows/x64/src/block/block_reverse_https.asm
@@ -145,7 +145,7 @@ download_more:
   test eax,eax           ; download failed? (optional?)
   jz failure

-  mov ax, word ptr [edi]
+  mov ax, word [edi]
   add rbx, rax           ; buffer += bytes_received

   test rax,rax           ; optional?
```

2) On `external/source/shellcode/windows/x86/src/build.py`:

```
$ python build.py all
...
src/stager/stager_reverse_winhttp.asm:17: fatal: unable to open include file `./src/block/block_reverse_winhttp_http.asm'
[-]  [Errno 2] No such file or directory: 'bin/stager_reverse_winhttp.bin'
```

Here is the change required to make this run:
```
$ git diff external/source/shellcode/windows/x86/src/stager/stager_reverse_winhttp.asm
diff --git a/external/source/shellcode/windows/x86/src/stager/stager_reverse_winhttp.asm b/external/source/shellcode/windows/x86/src/stager/stager_reverse_winhttp.asm
index 7e32a9b3b8..bb8af3e470 100644
--- a/external/source/shellcode/windows/x86/src/stager/stager_reverse_winhttp.asm
+++ b/external/source/shellcode/windows/x86/src/stager/stager_reverse_winhttp.asm
@@ -14,6 +14,6 @@
 %include "./src/block/block_api.asm"
 start:                   ;
   pop ebp                ; pop off the address of 'api_call' for calling later.
-%include "./src/block/block_reverse_winhttp_http.asm"
+%include "./src/block/block_reverse_winhttp.asm"
   ; By here we will have performed the reverse_tcp connection and EDI will be our socket.
```

Should I add that change to this PR? Similar to above, I diff-ed the results from master and this branch after the edit was made, and the results were identical.

I think the first checkbox can be checked.

If you could give me some pointers for testing the other ones that would be great.

**Comment by xmunoz** [comment]
ping @acammack-r7

**Comment by acammack-r7** [comment]
1. This is a NASM/MASM issue, and since it's Windows shellcode I'm inclined to leave it in the MASM style (IIRC MASM does require `ptr` in this case).
1. That does look like a bug in the assembly. This PR is works great, thanks!

In Metasploit the verification steps are mostly for the person landing the PR to verify all the proposed changes have their expected result. Verifying the rest of the changes will require a [Metasploit development environment](https://github.com/rapid7/metasploit-framework/wiki/Setting-Up-a-Metasploit-Development-Environment) and a working knowledge of how to use `msfconsole` (the [SANS cheat sheet](https://www.sans.org/security-resources/sec560/misc_tools_sheet_v1.pdf) is a little out of date, but has the basics). Each of those PRs has a list of verification steps like this list the software that needs to be installed to test the module and how the modules are expected to run. If you want to go through all that, that's great and will definitely help us get this landed sooner! Even if you do, though, someone else will still need to verify the changes you have pushed do the right thing (a sad result of us not having good automated testing around modules).

I'll try to set aside some time to look at this since I have most of this software set up already, but the code looks good and any the team members that want to take this on can.

**Comment by busterb** [comment]
Noting that a few of these don't work after the upgrade, will push fixes to the scripts.

**Comment by acammack-r7** [comment]
@busterb Did you the discussion the NASM/MASM difference?

> This is a NASM/MASM issue, and since it's Windows shellcode I'm inclined to leave it in the MASM style (IIRC MASM does require ptr in this case).

**Comment by busterb** [comment]
Ah, thanks, maybe can make them both happy. I'm suspect the last time this was built and updated statically in framework though, it wasn't with masm.

**Comment by acammack-r7** [comment]
That is a good point, looking again at it now it calls`nasm` explicitly. Removing `ptr` should be fine, then.

**Comment by busterb** [comment]
It might be possible to rewrite the asm on the fly so we filter keywords to make it happy in the build process, but I think just preferring nasm is the best case, since this is an area somewhat rarely visited by most users and developers.

File: tools/hardware/killerbee_msfrelay.py [hunk 2]
```diff
@@ -1,4 +1,4 @@
-#!/usr/bin/env python
```
Comment (busterb) [inline tools/hardware/killerbee_msfrelay.py]: there is no python3 release of killerbee, making this conversion not really possible at this time.

**Comment by busterb** [comment]
## Release Notes

This updates most Python code in Metasploit Framework to be compatible with Python 3.

**Pull request merged**